Add subPath e2e: test permissions and when subPath pre-exists

wongma7 opened the pull request:
These tests cover issues https://github.com/kubernetes/kubernetes/issues/45613 and ~~https://github.com/kubernetes/kubernetes/pull/43775~~ https://github.com/kubernetes/kubernetes/issues/41638
-->
```release-note
NONE
```

gnufied commented:
@k8s-bot ok to test

wongma7 commented:
I made an error in "should support existing directory subPath", fixing now.

jeffvance commented:
/assign jeffvance

jeffvance commented on test/e2e/common/host_path.go:
```diff
@@ -111,6 +112,71 @@ var _ = framework.KubeDescribe("HostPath", func() {
 		})
 	})
+
+	It("should support existing directory subPath [Volume]", func() {
```
existing directory or file, right?

wongma7 replied:
no, I wrote 1 for directory and 1 for file.

jeffvance commented on test/e2e/common/empty_dir.go:
```diff
@@ -153,6 +161,37 @@ func doTestSetgidFSGroup(f *framework.Framework, image string, medium v1.Storage
 
+func doTestSubPathFSGroup(f *framework.Framework, image string, medium v1.StorageMedium) {
+	var (
+		volumePath = "/test-volume"
```
consider making this a const usable by various tests. This string is assigned to `volumePath` 8 times.

jeffvance commented on test/e2e/common/host_path.go:
```diff
@@ -111,6 +112,71 @@ var _ = framework.KubeDescribe("HostPath", func() {
 	})
+
+	It("should support existing directory subPath [Volume]", func() {
+		volumePath := "/test-volume"
```
ditto const comment from above. 7 occurrences of this string.

jeffvance commented on test/e2e/common/host_path.go:
```diff
@@ -111,6 +112,71 @@ var _ = framework.KubeDescribe("HostPath", func() {
+
+	It("should support existing directory subPath [Volume]", func() {
+		volumePath := "/test-volume"
+		subPath := "sub-path"
```
Probably this should be a const too. 3 occurrences.

wongma7 replied:
I just chagned the value in "should support existing file subPath", idk if that test is too confusing now

wongma7 replied:
let me edit it, it indeed is way too confusing with subPath and fileName both.

jeffvance commented on test/e2e/common/host_path.go:
```diff
@@ -111,6 +112,71 @@ var _ = framework.KubeDescribe("HostPath", func() {
+		pod := testPodWithHostVol(volumePath, source)
+		// Create the subPath directory on the host
+		existing := path.Join(source.Path, subPath)
+		if err := os.Mkdir(path.Join(source.Path, subPath), 0755); err != nil {
```
More idiomatic as:
```
err := os.Mkdir(path.Join(source.Path, subPath), 0755))
Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("could not create directory %s", existing))
```

jeffvance commented on test/e2e/common/host_path.go:
```diff
@@ -111,6 +112,71 @@ var _ = framework.KubeDescribe("HostPath", func() {
+		pod := testPodWithHostVol(volumePath, source)
+		// Create the subPath file on the host
+		existing := path.Join(source.Path, fileName)
+		ioutil.WriteFile(existing, []byte("mount-tester new file"), 0644)
```
should check returned error

jeffvance commented:
@k8s-bot test this

jeffvance commented:
@k8s-bot bazel test this

saad-ali commented:
/lgtm
/approve

wongma7 commented:
something's wrong with the test cleanup, /tmp/test-file direcotry is sticking around from previous hostPath tests and making later ones fail. investigating.

wongma7 commented:
sigh it's because of the sticky bit on /tmp that the test cleanup doesn't actually work. It never actually mattered when tehre was only one test touching /tmp/test-file.

in the interest of getting this merged asap I'll just choose different filenames and add a TODO, the tests are failing because of something irrelevant to what they're testing...

Commit:
Add subPath e2e: test permissions and when subPath pre-exists

wongma7 commented:
@k8s-bot gce etcd3 e2e test this

jeffvance commented on test/e2e/common/host_path.go:
```diff
@@ -94,13 +92,95 @@ var _ = framework.KubeDescribe("HostPath", func() {
+		result, err := framework.SSH(fmt.Sprintf("mkdir -p %s", existing), framework.GetNodeExternalIP(&nodeList.Items[0]), framework.TestContext.Provider)
+		framework.ExpectNoError(err)
+		if result.Code != 0 {
+			framework.LogSSHResult(result)
```
I would log SSHResult independent of `Result.Code`

wongma7 replied:
done

jeffvance commented on test/e2e/common/host_path.go:
```diff
@@ -94,13 +92,95 @@ var _ = framework.KubeDescribe("HostPath", func() {
+		})
+	})
+
+	It("should support existing single file subPath [Volume]", func() {
```
There is a lot of common code in the 2 subPath `It` tests. Since there is some urgency in this pr it's not worth refactoring now. But a TODO would be to consolidate these It() blocks...

wongma7 replied:
added TODO comment

jeffvance commented on test/e2e/common/host_path.go:
```diff
@@ -94,13 +92,95 @@ var _ = framework.KubeDescribe("HostPath", func() {
+		result, err := framework.SSH(fmt.Sprintf("echo \"mount-tester new file\" > %s", existing), framework.GetNodeExternalIP(&nodeList.Items[0]), framework.TestContext.Provider)
+		framework.ExpectNoError(err)
+		if result.Code != 0 {
+			framework.LogSSHResult(result)
```
dittor on logging the result

wongma7 replied:
done

jeffvance commented on test/e2e/common/host_path.go:
```diff
@@ -94,13 +92,95 @@ var _ = framework.KubeDescribe("HostPath", func() {
+
+		// Create the subPath file on the host
+		existing := path.Join(source.Path, subPath)
+		result, err := framework.SSH(fmt.Sprintf("echo \"mount-tester new file\" > %s", existing), framework.GetNodeExternalIP(&nodeList.Items[0]), framework.TestContext.Provider)
```
I suppose you only need to `touch` the file vs. writing to it...

jeffvance replied:
Would be a bit better if the string written to the file and tested for when reading the file was a const.

wongma7 replied:
I thought so too, but the no need for \n at end of string here tripped me up; will leave this as TODO

wongma7 commented:
addressed jeff's second comments and added GetNodeExternalOrInternalIP in an attempt to make the tests work on "GCE Node e2e" in which nodes dont have external IPs

wongma7 commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

wongma7 commented:
@k8s-bot kops aws e2e test this

wongma7 commented:
Please re-add lgtm, (required) tests are now green and I'm  done making changes  @saad-ali 

Thanks 🙏

jeffvance commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

jeffvance commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

enisoc commented:
/lgtm
